### PR TITLE
[dtensor] add op support for aten.gather.default

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118917
* __->__ #118513

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

### CODE REVIEW

**Comment by XilunWu** [comment]
The CI report says "test_dtensor_op_db_take_along_dim_cpu_float32 in `test_dtensor_ops.py` has unexpected success". This means your change has made it working correctly. You can remove "xfail(take_along_dim)" from the file which will mark this test as a passing test instead of an expected failing test.

Another thing you can do in future is to locally run `pytest test/distributed/_tensor/test_dtensor_ops.py` to see if the test result has changed with your PR.

**Review by XilunWu — COMMENTED** [review]
left 2 suggestions. I'm not very clear about the main body of `gather_strategy` and the use of `_MaskPartial` though.

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 1]
```diff
@@ -94,7 +94,12 @@ def _reduce_value(
         assert self.mask_buffer.data is not None
 
         # apply the mask to the tensor that pending reduction
-        tensor[self.mask_buffer.data, :] = 0.0
+        # NOTE: Depending on the use case (gather op or embedding op),
+        # tensor dim can be equal or larger than the mask dim, respectively.
+        if tensor.ndim == self.mask_buffer.data.ndim:
+            tensor[self.mask_buffer.data] = 0.0
+        else:
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:101]: I think the main reason here is that the embedding output would produce an additional dimension compare to the input, hence the output masking logic become different? maybe have more clarification in the comment

File: torch/distributed/_tensor/ops/embedding_ops.py [hunk 2]
```diff
@@ -115,7 +120,12 @@ def _reduce_shard_value(
         assert self.mask_buffer.data is not None
 
         # apply the mask to the tensor that pending reduction
-        tensor[self.mask_buffer.data, :] = 0.0
+        # NOTE: Depending on the use case (gather op or embedding op),
+        # tensor dim can be equal or larger than the mask dim, respectively.
+        if tensor.ndim == self.mask_buffer.data.ndim:
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/embedding_ops.py:125]: given that we are reusing the logic, let's put this logic as a common method in MaskBuffer.

**Review by wanchaol — APPROVED** [review]
lgtm, some minor suggestions inlined.

**Comment by tianyu-l** [comment]
> The CI report says "test_dtensor_op_db_take_along_dim_cpu_float32 in `test_dtensor_ops.py` has unexpected success". This means your change has made it working correctly. You can remove "xfail(take_along_dim)" from the file which will mark this test as a passing test instead of an expected failing test.
> 
> Another thing you can do in future is to locally run `pytest test/distributed/_tensor/test_dtensor_ops.py` to see if the test result has changed with your PR.

This PR doesn't touch `take_along_dim`. Not sure how it could make the test pass... Should I still remove it?

File: torch/distributed/_tensor/ops/tensor_ops.py [hunk 3]
```diff
@@ -330,6 +334,70 @@ def replica_only_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType
...
 
 
+@register_op_strategy(aten.gather.default)
+def gather_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> StrategyType:
+    input_strategy = cast(OpStrategy, op_schema.args_schema[0])
+    dim = cast(int, op_schema.args_schema[1])
+    index_strategy = cast(OpStrategy, op_schema.args_schema[2])
+
+    input_shape = input_strategy.output_shape
+    index_shape = index_strategy.output_shape
```
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/_tensor/ops/tensor_ops.py:344]: If we don't check in dtensor sharding prop, these errors would pop up from local tensor ops, which seems still is the expected behavior?

**Comment by XilunWu** [comment]
> > The CI report says "test_dtensor_op_db_take_along_dim_cpu_float32 in `test_dtensor_ops.py` has unexpected success". This means your change has made it working correctly. You can remove "xfail(take_along_dim)" from the file which will mark this test as a passing test instead of an expected failing test.
> > Another thing you can do in future is to locally run `pytest test/distributed/_tensor/test_dtensor_ops.py` to see if the test result has changed with your PR.
> 
> This PR doesn't touch `take_along_dim`. Not sure how it could make the test pass... Should I still remove it?

@tianyu-l check https://github.com/pytorch/pytorch/blob/fb8ffba47ff0187cf3894cbe97aa5ad0a4739560/torch/_refs/__init__.py#L4470-L4504

**Review by XilunWu — APPROVED** [review]
LGTM! Thx for the work!

**Pull request closed without merging**